cocotb/handle.py

```diff
@@ -530,11 +519,52 @@ def _range_iter(self, left, right):
 
     @NonHierarchyObject.value.getter
     def value(self):
-        # need to iterate over the sub-object
-        result = []
-        for x in range(len(self)):
-            result.append(self[x].value)
-        return result
+        """A list of each value within this simulation object.
+
+        Getting and setting the current value of an array is done
+        by iterating through sub-handles in left-to-right order.
+
+        Given an HDL array ``arr``:
+
+        +--------------+---------------------+--------------------------------------------------------------+
+        | Verilog      | VHDL                | ``arr.value`` is equivalent to                               |
+        +==============+=====================+==============================================================+
+        | ``arr[4:6]`` | ``arr(4 to 7)``     | ``[arr[4].value, arr[5].value, arr[6].value, arr[7].value]`` |
+        +--------------+---------------------+--------------------------------------------------------------+
+        | ``arr[7:4]`` | ``arr(7 downto 4)`` | ``[arr[7].value, arr[6].value, arr[5].value, arr[4].value]`` |
+        +--------------+---------------------+--------------------------------------------------------------+
+
+        When setting this property as in ``arr.value = ...``, the same index equivalence as noted in the table holds.
+
+        .. note::
+            When setting this property, the values will be cached as explained in :attr:`ModifiableObject.value`.
+
+        .. warning::
+            Assigning a value to a sub-handle:
+
+            - **Wrong**: ``dut.some_array.value[0] = 1`` (Gets value as a list then updates index 0)
+            - **Correct**: ``dut.some_array[0].value = 1``
+        """
+        # Don't use self.__iter__, because it has an unwanted `except IndexError`
+        return [
+            self[i]
+            for i in self._range_iter(self._range[0], self._range[1])
+        ]
+
+    @value.setter
+    def value(self, value):
+        """Assign value from a list of same length to an array in left-to-right order.
+        Index 0 of the list maps to the left-most index in the array.
+
+        See the docstring for :attr:`value` above.
+        """
+        if type(value) is not list:
+            raise TypeError("Assigning non-list value to object %s of type %s" % (self._name, type(self)))
+        if len(value) != len(self):
+            raise IndexError("Assigning list of length %d to object %s of length %d" % (
+                len(value), self._name, len(self)))
+        for val_idx, self_idx in enumerate(self._range_iter(self._range[0], self._range[1])):
+            self[self_idx] = value[val_idx]
```

Comment:
Should this be:
```suggestion
            self[self_idx].value = value[val_idx]
```
Otherwise it goes through `__setitem__`, which resolves to the same thing once I fix it per other comment.

Reply:
It doesn't make much difference. My reading of your comment if that you're worried about recursion here, but you needn't be.

Having said that, using `.value` will avoid some indirection and be slightly faster.